t/gm.t

```diff
@@ -128,3 +151,221 @@ New, NTLSv1.1, Cipher is ECDHE-SM2-SM4-CBC-SM3
+        }
+        ssl_certificate ../../certs/apisix.crt;
+        ssl_certificate_key ../../certs/apisix.key;
+
+        server_tokens off;
+        location / {
+            content_by_lua_block {
+                ngx.shared.done:set("handshake", true)
+            }
+        }
+    }
+--- config
+    server_tokens off;
+
+    location /t {
+        content_by_lua_block {
+            ngx.shared.done:delete("handshake")
+            local f, err = handshake()
+            if not f then
+                ngx.say(err)
+                return
+            end
+
+            local step = 0.001
+            while step < 1 do
+                ngx.sleep(step)
+                step = step * 2
+
+                if ngx.shared.done:get("handshake") then
+                    local out = f:read('*a')
+                    ngx.log(ngx.INFO, out)
+                    ngx.say("ok")
+                    f:close()
+                    return
+                end
+            end
+        }
+    }
+
+--- error_log
+SSL_do_handshake() failed
+--- no_error_log
+[error]
+[alert]
+[emerg]
+
+
+
+=== TEST 3: gm handshake, then disable ntls
+--- http_config
+    init_worker_by_lua_block {
+        local ssl = require "resty.apisix.ssl"
+        ssl.enable_ntls()
+    }
+
+    lua_shared_dict done 16k;
+    server {
+        listen 1994 ssl;
+        server_name   test.com;
+        ssl_certificate_by_lua_block {
+            set_gm_cert_and_key()
+        }
+        ssl_certificate ../../certs/apisix.crt;
+        ssl_certificate_key ../../certs/apisix.key;
+
+        server_tokens off;
+        location / {
+            content_by_lua_block {
+                ngx.shared.done:set("handshake", true)
+            }
+        }
+    }
+--- config
+    server_tokens off;
+
+    location /t {
+        content_by_lua_block {
+            local ssl = require "resty.apisix.ssl"
+            ssl.disable_ntls()
+
+            ngx.shared.done:delete("handshake")
+            local f, err = handshake()
+            if not f then
+                ngx.say(err)
+                return
+            end
+
+            local step = 0.001
+            while step < 2 do
+                ngx.sleep(step)
+                step = step * 2
+
+                if ngx.shared.done:get("handshake") then
+                    local out = f:read('*a')
+                    ngx.log(ngx.INFO, out)
+                    ngx.say("ok")
+                    f:close()
+                    return
+                end
+            end
+        }
+    }
+
+--- error_log
+SSL_do_handshake() failed
+--- no_error_log
+[error]
+[alert]
+[emerg]
+
+
+
+=== TEST 4: regular handshake with gm enabled
+--- http_config
+    init_worker_by_lua_block {
+        local ssl = require "resty.apisix.ssl"
+        ssl.enable_ntls()
+    }
+
+    lua_shared_dict done 16k;
+    server {
+        listen 1994 ssl;
+        server_name   test.com;
+        ssl_certificate_by_lua_block {
+            local ngx_ssl = require "ngx.ssl"
```

Comment:
Reuse `set_gm_cert_and_key ` ?

Reply:
The way to set GM cert is too different to reuse. They only have few similarities.